### PR TITLE
Allow keyboard input for keypad

### PR DESCRIPTION
**Pull request by FloEdelmann**
Until now, one had to click the buttons in the web interface to control the keypad. I've enabled keyboard input for this.

### CODE REVIEW

File: olad/www/new/js/app.js [hunk 1]
```diff
@@ -310,9 +310,56 @@ ola.controller('keypadUniverseCtrl',
...
+          case '7':
+          case '8':
+          case '9':
+            $scope.input(key);
+            break;
+          case '@':
+          case 'a':
+            $scope.input(' @ ');
+            break;
+          case '-':
```
Comment (peternewman) [inline olad/www/new/js/app.js]: I'd be inclined to leave out -, so we can do  1 THRU 10 - 5 + 15 in future, without confusing things.

**Comment by peternewman** [comment]
Thanks for this!

File: olad/www/new/js/app.js (hunk 1 above)
Comment (peternewman) [inline olad/www/new/js/app.js]: It was popular on Strand, and probably other consoles. The above would select channels 1, 2, 3, 4, 6, 7, 8, 9, 10 and 15.

**Comment by FloEdelmann** [comment]
Should the directive also go into the file you mentioned?

File: olad/www/new/js/app.js (hunk 1 above)
Comment (peternewman) [inline olad/www/new/js/app.js]: What else would you expect it to do (aside from maths)?

**Comment by peternewman** [comment]
Possibly a question for @daveol but from my understanding of directives and given what he did with filters, I'd say make a new folder and put it in there, similar to this:
https://github.com/OpenLightingProject/ola/tree/master/javascript/new-src/src/filters

I'm assuming that directive can be used to grab focus anywhere in the UI we might need to?

**Comment by peternewman** [comment]
Yeah * did all channels before (min channel to max channel).

**Comment by peternewman** [comment]
Travis is now working again, there are a chunk of JS lint issues you'll need to fix please:
https://travis-ci.org/OpenLightingProject/ola/jobs/169594371#L417

**Comment by FloEdelmann** [comment]
jshint and jscs are don't agree on the case indentation. When I comply with one, the other fails.

**Comment by peternewman** [comment]
I'm inclined to say we should stick with jscs as you've currently done, as that's our own custom JS style, rather than a default one. So then it's just a case of stopping jshint complaining.

It looks like this got resolved in jshint here, which should be in jshint 2.2.0 upwards:
https://github.com/jshint/jshint/commit/c64fb5c680f67538b095e496a56f48e79ee670ec

Which we're already running, so needs a little more digging.

This is the first switch/case in our JS code, hence the issue.

**Comment by peternewman** [comment]
Actually I think it may be the {} block within the case @FloEdelmann . Have you tried removing that?

**Comment by FloEdelmann** [comment]
I actually never used `{}` blocks for case statements but added them after jscs complained.

**Comment by peternewman** [comment]
I've taken the {} back out, as we need to fix our JCSC config instead.

**Comment by peternewman** [comment]
Travis should be fixed by https://github.com/OpenLightingProject/javascript-style/pull/9 .

**Comment by peternewman** [comment]
As promised, now all green.

File: javascript/new-src/src/directives/autofocus.js [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Library General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Allows an element to obtail focus whenever a variable is set to true.
```
Comment (peternewman) [inline javascript/new-src/src/directives/autofocus.js]: Fixed.

File: javascript/new-src/src/controllers/keypad_universe.js [hunk 3]
```diff
@@ -79,9 +79,61 @@ ola.controller('keypadUniverseCtrl',
...
           $scope.field = fields[0];
         }
+        $scope.focusInput = true;
+      };
+
+      $scope.keypress = function($event) {
+        var key = $event.key;
+
+        if ($event.altKey || $event.ctrlKey || $event.metaKey
+            || ($event.which == 0 && key != 'Enter' && key != 'Backspace')) {
```
Comment (peternewman) [inline javascript/new-src/src/controllers/keypad_universe.js]: Sorted

File: javascript/new-src/src/directives/autofocus.js (hunk 2 above)
Comment (peternewman) [inline javascript/new-src/src/directives/autofocus.js]: Fixed

**Comment by FloEdelmann** [comment]
Nice, thank you :)

**Comment by FloEdelmann** [comment]
Exactly. Ok I'll do this then.


Am 21.10.2016 um 18:36 schrieb Peter Newman:

**Comment by peternewman** [comment]
@daveol if you want to retrospectively review this, let us know if there are any issues.

@FloEdelmann I realised I forgot to regenerate the JS after fixing lint issues and it looks like you did the same. Would you mind updating them and pushing a new PR please.

**Comment by FloEdelmann** [comment]
Unfortunately, I won't find the time until maybe Monday. If it's still relevant then, just let me know.

**Pull request merged**